### DISCUSSION TITLE
java.util.date java.lang.UnsupportedOperationException: class redefinition failed: attempted to change superclass or interfaces

### DISCUSSION THREAD

**Discussion by RainNiu** [apache/skywalking#10692] [Q&A] [answered]
I want to use an agent to rewrite new Date() and then rewrite the system time of the current thread of the application. The application only uses one Java agent.This problem similar issues #9701 , but I don't see how to solve it. Please help me. thanks.

How to reproduce:

Agent version:8.15.0
Plugin：jdk-date-plugin
The command is like '-javaagent:skywalking-agent.jar'.  

Error Message:
INFO 2023-04-14 20:54:48.345 main PluginResourcesResolver : find skywalking plugin define in jar:file:/Users/Shared/ProjectFiles/03-Workspace/GitHub/skywalking/skywalking-java/skywalking-agent/plugins/apm-jdk-threadpool-plugin-8.15.0.jar!/skywalking-plugin.def 
DEBUG 2023-04-14 20:54:48.350 main PluginBootstrap : loading plugin class org.apache.skywalking.apm.plugin.define.DateInstrumentation. 
DEBUG 2023-04-14 20:54:48.362 main PluginBootstrap : loading plugin class org.apache.skywalking.apm.plugin.define.ThreadPoolExecutorInstrumentation. 
ResetDateInstrumentation.enhanceClass
ResetDateInstrumentation.getInstanceMethodsInterceptPoints
ResetDateInstrumentation.getInstanceMethodsInterceptPoints
ResetDateInstrumentation.getConstructorsInterceptPoints
ResetDateInstrumentation.getConstructorsInterceptPoints
ResetDateInstrumentation.getStaticMethodsInterceptPoints
ResetDateInstrumentation.getStaticMethodsInterceptPoints
DEBUG 2023-04-14 20:55:30.782 main AbstractClassEnhancePluginDefine : prepare to enhance class java.util.Date by org.apache.skywalking.apm.plugin.define.DateInstrumentation. 
ResetDateInstrumentation.getStaticMethodsInterceptPoints
ResetDateInstrumentation.getConstructorsInterceptPoints
ResetDateInstrumentation.getInstanceMethodsInterceptPoints
DEBUG 2023-04-14 20:55:30.945 main AbstractClassEnhancePluginDefine : enhance class java.util.Date by org.apache.skywalking.apm.plugin.define.DateInstrumentation completely. 
DEBUG 2023-04-14 20:55:30.946 main SkyWalkingAgent : Finish the prepare stage for java.util.Date. 
DEBUG 2023-04-14 20:55:31.184 main SkyWalkingAgent : On Transformation class java.util.Date. 
ERROR 2023-04-14 20:55:37.428 main SkyWalkingAgent : index=0, batch=[class java.util.Date], types=[class java.util.Date] 
java.lang.UnsupportedOperationException: class redefinition failed: attempted to change superclass or interfaces
	at sun.instrument.InstrumentationImpl.retransformClasses0(Native Method)
	at sun.instrument.InstrumentationImpl.retransformClasses(InstrumentationImpl.java:144)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at net.bytebuddy.utility.Invoker$Dispatcher.invoke(Unknown Source)
	at net.bytebuddy.utility.dispatcher.JavaDispatcher$Dispatcher$ForNonStaticMethod.invoke(JavaDispatcher.java:1032)
	at net.bytebuddy.utility.dispatcher.JavaDispatcher$ProxiedInvocationHandler.invoke(JavaDispatcher.java:1162)
	at net.bytebuddy.agent.builder.$Proxy13.retransformClasses(Unknown Source)
	at net.bytebuddy.agent.builder.AgentBuilder$RedefinitionStrategy$Collector$ForRetransformation.doApply(AgentBuilder.java:8177)
	at net.bytebuddy.agent.builder.AgentBuilder$RedefinitionStrategy$Collector.apply(AgentBuilder.java:7992)
	at net.bytebuddy.agent.builder.AgentBuilder$RedefinitionStrategy.apply(AgentBuilder.java:5689)
	at net.bytebuddy.agent.builder.AgentBuilder$Default.doInstall(AgentBuilder.java:11126)
	at net.bytebuddy.agent.builder.AgentBuilder$Default.installOn(AgentBuilder.java:11044)
	at org.apache.skywalking.apm.agent.SkyWalkingAgent.premain(SkyWalkingAgent.java:136)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at sun.instrument.InstrumentationImpl.loadClassAndStartAgent(InstrumentationImpl.java:386)
	at sun.instrument.InstrumentationImpl.loadClassAndCallPremain(InstrumentationImpl.java:401)

<img width="1358" alt="image" src="https://user-images.githubusercontent.com/23009238/232053755-cbb2f6a9-06c0-4586-a1fa-04964215afd4.png">

My Code:

package org.apache.skywalking.apm.plugin.define;

import net.bytebuddy.description.method.MethodDescription;
import net.bytebuddy.matcher.ElementMatcher;
import net.bytebuddy.matcher.ElementMatchers;
import org.apache.skywalking.apm.agent.core.plugin.interceptor.ConstructorInterceptPoint;
import org.apache.skywalking.apm.agent.core.plugin.interceptor.InstanceMethodsInterceptPoint;
import org.apache.skywalking.apm.agent.core.plugin.interceptor.StaticMethodsInterceptPoint;
import org.apache.skywalking.apm.agent.core.plugin.interceptor.enhance.ClassInstanceMethodsEnhancePluginDefine;
import org.apache.skywalking.apm.agent.core.plugin.match.ClassMatch;
import org.apache.skywalking.apm.agent.core.plugin.match.NameMatch;

public class DateInstrumentation extends ClassInstanceMethodsEnhancePluginDefine {

    private static final String ENHANCE_CLASS = "java.util.Date";

    private static final String RESETDATE_INTERCEPTOR = "org.apache.skywalking.apm.plugin.DateInterceptor";

    @Override
    public boolean isBootstrapInstrumentation() {
        return true;
    }

    @Override
    protected ClassMatch enhanceClass() {
        System.out.println("ResetDateInstrumentation.enhanceClass");
        // 指定想要监控的类
        return NameMatch.byName(ENHANCE_CLASS);
    }

    @Override
    public ConstructorInterceptPoint[] getConstructorsInterceptPoints() {
        System.out.println("ResetDateInstrumentation.getConstructorsInterceptPoints");
//        return new ConstructorInterceptPoint[0];
        return new ConstructorInterceptPoint[]{
                new ConstructorInterceptPoint() {
                    @Override
                    public ElementMatcher<MethodDescription> getConstructorMatcher() {
                        return ElementMatchers.takesArguments(1);
                    }

                    @Override
                    public String getConstructorInterceptor() {
                        return RESETDATE_INTERCEPTOR;
                    }
                }
        };
    }

    @Override
    public InstanceMethodsInterceptPoint[] getInstanceMethodsInterceptPoints() {
        // 指定想要监控的实例方法，每个实例方法对应一个InstanceMethodsInterceptPoint
        System.out.println("ResetDateInstrumentation.getInstanceMethodsInterceptPoints");
        return new InstanceMethodsInterceptPoint[0];
    }

    @Override
    public StaticMethodsInterceptPoint[] getStaticMethodsInterceptPoints() {
        System.out.println("ResetDateInstrumentation.getStaticMethodsInterceptPoints");
        return new StaticMethodsInterceptPoint[0];
    }

}

**Comment by wu-sheng** [comment, marked as answer]
Generally, this is a question out of SkyWalking scope. 
I don't think it is suitable for us to follow up.
Date could be never a class required to be instrumented. Don't consider the SkyWalking agent is a general-use agent. Byte-buddy(core lib used by SkyWalking) should be a lib you dive into.

**Comment by wu-sheng** [comment]
> java.lang.UnsupportedOperationException: class redefinition failed: attempted to change superclass or interfaces

This is a very binary code-level error. Usually, it means you are instrumenting the wrong method. 
Which it is and why are belonging to another topic? This is what I mean out of our scope too.
